### PR TITLE
A02 Ilia Malakhov

### PR DESCRIPTION
**Pull request by ilma4**

### CODE REVIEW

File: src/X86.lama [hunk 1]
```diff
@@ -283,23 +283,60 @@ fun compile (env, code) {
...
+    code := code <+ Mov(a, eax);
+    case s of 
+      "+" -> code <+ Binop(s, b, eax) <+ Mov(eax, res)
+    | "-" -> code <+ Binop(s, b, eax) <+ Mov(eax, res)
+    | "*" -> code <+ Binop(s, b, eax) <+ Mov(eax, res)
+
+    | "/" -> code <+ IDiv(b) <+ Mov(eax, res)
+    | "%" -> code <+ IDiv(b) <+ Mov(edx, res)
+
+    | "&&" -> code <+ Binop("!!", b, eax) <+ Set("ne", "%al")<+ Binop("&&", L(1), eax) <+ Mov(eax, res)
```
Comment (e2e4b6b7) [inline src/X86.lama:327]: 1 && 2 will result false as far as I understood this code. This is incorrect, you have to normalize the int value

File: src/X86.lama [hunk 2]
```diff
@@ -283,23 +283,60 @@ fun compile (env, code) {
...
+    | "<" -> code <+ Binop("cmp", b, eax) <+ Set("l", "%al") <+ Binop("&&", L(1), eax) <+ Mov(eax, res)
+    | ">" -> code <+ Binop("cmp", b, eax) <+ Set("g", "%al") <+ Binop("&&", L(1), eax) <+ Mov(eax, res)
+    | "<=" -> code <+ Binop("cmp", b, eax) <+ Set("le", "%al") <+ Binop("&&", L(1), eax) <+ Mov(eax, res)
+    | ">=" -> code <+ Binop("cmp", b, eax) <+ Set("ge", "%al") <+ Binop("&&", L(1), eax) <+ Mov(eax, res)
+    | "==" -> code <+ Binop("cmp", b, eax) <+ Set("e", "%al") <+ Binop("&&", L(1), eax) <+ Mov(eax, res)
+    | "!=" -> code <+ Binop("cmp", b, eax) <+ Set("ne", "%al") <+ Binop("&&", L(1), eax) <+ Mov(eax, res)
+    | _ -> failure("codegeneration for BINOP %s is not yet implemented\n", s.string)
```
Comment (e2e4b6b7) [inline src/X86.lama:334]: where is `!!`?

**Review by e2e4b6b7 — CHANGES_REQUESTED** [review]

File: src/X86.lama [hunk 3]
```diff
@@ -283,23 +283,60 @@ fun compile (env, code) {
...
       | _ -> failure ("codegeneration for instruction %s is not yet implemented\n", i.string)
       esac
     }, [env, emptyBuffer ()], code)
 }
 
+fun compileBinop([env, scode], s) {
+  -- printf("Compiling binop: %s. env: %s. code: %s", s.string, env.string, code.string);
+  var code = scode;
+  case pop2(env) of [b, a, env] -> case allocate(env) of [res, env] -> [env,
+    code := code <+ Mov(a, eax);
```
Comment (e2e4b6b7) [inline src/X86.lama:318]: Sorry for spamming. This instruction is redundant for most of cases (imperformant)

File: src/X86.lama (hunk 1 above)
Comment (ilma4) [reply src/X86.lama]: 1 && 2 will result true as expected. The issue is 1 && 0 will result true as well.
Fixed in the next commits

File: src/SM.lama [hunk 4]
```diff
@@ -32,16 +32,22 @@ fun eval (cfg, insns) {
   foldl (fun ([stack, state, world], instr){
     -- printf("Executing instrutions: %s\n", instr.string);
     case instr of
-        READ -> case readWorld(world) of [n, newWorld] -> [n:stack, state, newWorld] esac
+        READ -> case readWorld(world) of 
+            [n, world] -> [n:stack, state, world] 
+          | _ -> failure("Error in SM.eval: can't read from world while READ") esac
```
Comment (e2e4b6b7) [inline src/SM.lama:37]: nit: Actually, this case is impossible as the function return pair

File: src/X86.lama [hunk 5]
```diff
@@ -311,30 +313,83 @@ fun compile (env, code) {
...
-    | ">=" -> code <+ Binop("cmp", b, eax) <+ Set("ge", "%al") <+ Binop("&&", L(1), eax) <+ Mov(eax, res)
-    | "==" -> code <+ Binop("cmp", b, eax) <+ Set("e", "%al") <+ Binop("&&", L(1), eax) <+ Mov(eax, res)
-    | "!=" -> code <+ Binop("cmp", b, eax) <+ Set("ne", "%al") <+ Binop("&&", L(1), eax) <+ Mov(eax, res)
-    | _ -> failure("codegeneration for BINOP %s is not yet implemented\n", s.string)
+  case pop2(env) of [b, a, env] -> [push(env, a),
+    case op of 
+      "+" -> code <+> compilePm(op, b, a)
+    | "-" -> code <+> compilePm(op, b, a)
+
+    | "*" -> code <+ Mov(a, eax) <+ Binop(op, b, eax) <+ Mov(eax, a)
```
Comment (e2e4b6b7) [inline src/X86.lama:323]: nit: Still redundant move if a and b both in registers

File: src/X86.lama [hunk 6]
```diff
@@ -311,30 +313,83 @@ fun compile (env, code) {
...
+  var code = emptyBuffer() <+ Mov(a, eax) <+ Cltd <+ IDiv(b);
+  case op of
+    "/" -> code <+ Mov(eax, a)
+  | "%" -> code <+ Mov(edx, a)
+  esac
+}
+
+fun compileOr(b, a){
+  if isRegister(a) !! isRegister(b) then 
+    emptyBuffer() <+ Binop("^", eax, eax) <+ Binop("!!", b, a) <+ Set("ne", "%al") <+ Mov(eax, a)
```
Comment (e2e4b6b7) [inline src/X86.lama:368]: It works but not follows the specification. The resulting value have to be zero or one AFAIK, not an arbitrary integer

File: src/X86.lama [hunk 7]
```diff
@@ -311,30 +313,83 @@ fun compile (env, code) {
...
+  else
+    emptyBuffer() <+ Mov(a, eax) <+ Binop("!!",  b, eax) <+ Set("ne", "%al") <+ Binop("&&", L(1), eax) <+ Mov(eax, a)
+  fi
+}
+
+fun compileAnd(b, a){
+  emptyBuffer() <+ Binop("^", eax, eax) <+ cmpWithZeroInsn(a) <+ Set("ne", "%al")
```
Comment (e2e4b6b7) [inline src/X86.lama:375]: edx is not cleared but used as "%dl"

File: src/X86.lama [hunk 8]
```diff
@@ -311,30 +313,83 @@ fun compile (env, code) {
...
+  else
+    emptyBuffer() <+ Mov(a, eax) <+ Binop(op, b, eax) <+ Mov(eax, a)
+  fi
+}
+
+fun compileCompare(op, b, a) {
+  if isRegister(a) !! isRegister(b) then
+    emptyBuffer() <+ Binop("^", eax, eax) <+ Binop("cmp", b, a) <+ Set(suffix(op), "%al") <+ Mov(eax, a)
+  else
+    emptyBuffer() <+ Mov(a, eax) <+ Binop("cmp", b, eax) <+ Set(suffix(op), "%al") <+ Binop("&&", L(1), eax) <+ Mov(eax, a)
```
Comment (e2e4b6b7) [inline src/X86.lama:391]: This does not work. You have `a` in eax and then assigning result to the lowest bits of "%al". But you still have something unexpected in the highest bits

**Review by e2e4b6b7 — CHANGES_REQUESTED** [review]

File: src/X86.lama (hunk 6 above)
Comment (ilma4) [reply src/X86.lama:368]: I believe, it follows. First, eax is zeroed. Then "setne" sets the least byte to zero or one (the greater bits to zero and the least bit to 1/0). So it is exactry zero or one

File: src/X86.lama (hunk 8 above)
Comment (ilma4) [reply src/X86.lama:391]: So I clear it with bitwise && next operation, but only higher bits

**Comment by e2e4b6b7** [comment]
Okey, you are right in all comments. It is quite unique solution, but correct as far as I understand

**Pull request closed without merging**